### PR TITLE
Cranelift: Avoid calling `ensure_struct_return_pointer_is_returned` and cloning sigs for every call

### PR DESCRIPTION
**Pull request by fitzgen**
Unfortunately, this seems to be perf neutral. I think we already got most of the wins we otherwise would have got when we deduplicated IR signatures during lowering.

### CODE REVIEW

File: cranelift/codegen/src/machinst/abi.rs [hunk 1]
```diff
@@ -1173,19 +1176,21 @@ fn gen_store_stack_multi<M: ABIMachineSpec>(
...
-    let params_structret = sig
-        .params
-        .iter()
-        .find(|p| p.purpose == ArgumentPurpose::StructReturn);
-    let rets_have_structret = sig.returns.len() > 0
-        && sig
-            .returns
-            .iter()
-            .any(|arg| arg.purpose == ArgumentPurpose::StructReturn);
+/// If the signature needs to be legalized, the return the struct-return
```
Comment (jameysharp) [inline cranelift/codegen/src/machinst/abi.rs:1179]:
```suggestion
/// If the signature needs to be legalized, then return the struct-return
```

File: cranelift/codegen/src/machinst/abi.rs [hunk 2]
```diff
@@ -1173,19 +1176,21 @@ fn gen_store_stack_multi<M: ABIMachineSpec>(
...
+/// parameter that should be prepended to its returns. Otherwise, return `None`.
+fn missing_struct_return(sig: &ir::Signature) -> Option<ir::AbiParam> {
+    let struct_ret_index = sig.special_param_index(ArgumentPurpose::StructReturn)?;
+    if !(sig.returns.len() > 0 && sig.uses_special_return(ArgumentPurpose::StructReturn)) {
```
Comment (jameysharp) [inline cranelift/codegen/src/machinst/abi.rs:1183]: I haven't looked at the implementation of `uses_special_return`. Why does it need to be guarded to only run when there are some return values?

Maybe the answer is just that the old implementation had this guard. But I think the old implementation, using `Iterator::any`, shouldn't have needed this either: `any` returns `false` on an empty iterator.

**Review by jameysharp — APPROVED** [review]
I've suggested a couple minor changes, but this makes sense to me!

**Comment by fitzgen** [comment]
Actually, all my benchmarking had previously been done while disabling parallel compilation, and when we enable parallel compilation I see a consisten 2% speed up to compilation with this patch!

```
compilation :: instructions-retired :: benchmarks/pulldown-cmark/benchmark.wasm

  Δ = 157429.67 ± 7979.61 (confidence = 99%)

  no-legalize-sigs.so is 1.02x to 1.02x faster than main.so!

  [7087159 7141950.42 7197962] main.so
  [6938212 6984520.75 7045311] no-legalize-sigs.so

compilation :: instructions-retired :: benchmarks/spidermonkey/benchmark.wasm

  Δ = 3586422.00 ± 367931.21 (confidence = 99%)

  no-legalize-sigs.so is 1.02x to 1.02x faster than main.so!

  [169418264 172497906.79 173415861] main.so
  [165471107 168911484.79 169928589] no-legalize-sigs.so

compilation :: instructions-retired :: benchmarks/bz2/benchmark.wasm

  Δ = 35887.48 ± 5190.24 (confidence = 99%)

  no-legalize-sigs.so is 1.02x to 1.02x faster than main.so!

  [1867690 1896927.22 1937605] main.so
  [1834324 1861039.74 1893752] no-legalize-sigs.so
```

File: cranelift/codegen/src/machinst/abi.rs (hunk 2 above)
Comment (fitzgen) [reply cranelift/codegen/src/machinst/abi.rs:1183]: It isn't that `uses_special_return` forces this, it is the pre-existing logic of when we need to legalize, and I have no idea why it is like that, I just faithfully kept it the same.

Comment (cfallin) [reply cranelift/codegen/src/machinst/abi.rs]: Yeah, it looks like it was an over-eager guard that wasn't actually necessary. Latest version looks good.

**Review by cfallin — APPROVED** [review]
LGTM, nice!

**Pull request merged**